rs/index/src/posting_list/file.rs

```diff
@@ -177,60 +180,45 @@ impl FileBackedAppendablePostingListStorage {
             return Err(anyhow!("Posting lists should already be flushed to disk"));
         }
         // First write the metadata
-        if self.current_offset == self.backing_file_size {
-            self.new_backing_file()?;
+        if self.metadata_current_overall_offset % self.backing_file_size == 0 {
+            self.new_backing_file(true)?;
         }
         // Write the length of the posting list
-        self.write_to_current_mmap(&posting_list.len().to_le_bytes())?;
+        self.write_to_current_mmap(&posting_list.len().to_le_bytes(), true)?;
 
         // Write the offset to the current posting list
-        self.write_to_current_mmap(&self.offset_to_current_posting_list.to_le_bytes())?;
+        self.write_to_current_mmap(&self.current_overall_offset.to_le_bytes(), true)?;
 
         // Now write the posting list
-        let size_in_bytes = std::mem::size_of::<u64>();
-        let metadata_tot_size = self.get_posting_list_offset();
-        let file_num_for_first_posting_list = metadata_tot_size / self.backing_file_size;
-        // file_num_for_first_posting_list is an index, while self.mmaps.len() is the length
-        while self.mmaps.len() - 1 < file_num_for_first_posting_list {
-            self.new_backing_file()?;
-        }
-        let file_access_info =
-            self.offset_to_file_access_info(self.offset_to_current_posting_list as usize)?;
-        self.current_offset = file_access_info.file_offset;
-        self.current_backing_id = file_access_info.file_num as i32;
-
-        self.offset_to_current_posting_list += (posting_list.len() * size_in_bytes) as u64;
-
         for idx in posting_list.iter() {
-            if self.current_offset == self.backing_file_size {
-                self.new_backing_file()?;
+            if self.current_overall_offset % self.backing_file_size == 0 {
+                self.new_backing_file(false)?;
             }
-            self.write_to_current_mmap(&idx.to_le_bytes())?;
-        }
-
-        if self.current_offset == self.backing_file_size {
-            self.new_backing_file()?;
+            self.write_to_current_mmap(&idx.to_le_bytes(), false)?;
         }
 
-        self.current_num_of_posting_list += 1;
+        self.entry_count += 1;
         Ok(())
     }
 
-    fn offset_to_file_access_info(&self, offset: usize) -> Result<FileAccessInfo> {
-        let file_num = offset / self.backing_file_size;
-        if file_num >= self.mmaps.len() {
+    fn offset_to_file_access_info(
+        offset: usize,
+        file_size: usize,
+        mmaps: &[memmap2::MmapMut],
+    ) -> Result<FileAccessInfo> {
+        let file_num = offset / file_size;
+        if file_num >= mmaps.len() {
             return Err(anyhow!("File number out of bound"));
         }
 
-        let file_offset = offset % self.backing_file_size;
-        if file_offset >= self.mmaps[file_num].len() {
+        let file_offset = offset % file_size;
+        if file_offset >= mmaps[file_num].len() {
```

Comment:
dumb question: isn't `mmaps[file_num].len() == file_size`?

Reply:
No, that's only the case when `mmaps[file_num]` is full.